src/cryptonote_core/blockchain.cpp

```diff
@@ -3298,6 +3299,71 @@ bool Blockchain::expand_transaction_2(transaction &tx, const crypto::hash &tx_pr
+  // positives (and thus possibly enabling double spends) is we take a hash of two things. One,
+  // we use cryptonote::get_transaction_hash() which gives us a (cryptographically secure)
+  // unique representation of all "knobs" controlled by the possibly malicious constructor of
+  // the transaction. Two, we take a hash of all *previously validated* blockchain data
+  // referenced by this transaction which is required to validate the ring signature. In our
+  // case, this is the mixring. Future versions of the protocol may differ in this regard, but
+  // if this assumptions holds true in the future, enable the RCT cache by modifying the first
+  // two asserts, and the `rv.type != RCTTypeBulletproofPlus` "if" branch below.
+  CHECK_AND_ASSERT_MES(tx.version <= 2, false, "Unknown TX type. Make sure RCT cache works correctly with this type and then enable it in the code here.");
+  CHECK_AND_ASSERT_MES(rv.type <= rct::RCTTypeBulletproofPlus, false, "Unknown RCT type. Make sure RCT cache works correctly with this type and then enable it in the code here.");
+  CHECK_AND_ASSERT_MES(!tx.pruned, false, "Pruned transaction will not pass verRctNonSemanticsSimple");
+
+  // Do RCT expansion + full non-semantics verification. Has no captures so should optimize well
+  const auto expand_tx_and_ver_rct_non_sem =
+  [] (
+    cryptonote::transaction& t, // =tx
+    const crypto::hash& ph,     // =tx_prefix_hash
+    const rct::ctkeyM& mr       // =mix_ring
+  ) -> bool
+  {
+    const bool expand_res = cryptonote::Blockchain::expand_transaction_2(t, ph, mr);
+    CHECK_AND_ASSERT_MES(expand_res, false, "Failed to expand rct signatures!");
+    return verRctNonSemanticsSimple(t.rct_signatures);
+  };
+
+  // Don't cache older (or newer) rctSig types
+  // This cache only makes sense when it caches data from mempool first,
+  // so only "current fork version-enabled" RCT types need to be cached
+  if (rv.type != rct::RCTTypeBulletproofPlus)
+    return expand_tx_and_ver_rct_non_sem(tx, tx_prefix_hash, mix_ring);
+
+  // We want to create a domain seperated blob containing the full TX hash and mix_ring blob
+  std::stringstream ss;
+
+  // Start with domain seperation
+  ss << config::HASH_KEY_TXHASH_AND_MIXRING;
+
+  // Then add TX hash
+  const crypto::hash tx_hash = cryptonote::get_transaction_hash(tx);
+  ss.write(tx_hash.data, sizeof(crypto::hash));
+
+  // Then serialize mix ring
+  binary_archive<true> ar(ss);
+  ::do_serialize(ar, const_cast<rct::ctkeyM&>(mix_ring));
+
+  // Calculate hash of TX hash and mix ring blob
+  crypto::hash tx_and_mixring_hash;
+  cryptonote::get_blob_hash(ss.str(), tx_and_mixring_hash);
+
+  static tools::data_cache<crypto::hash, 8192> cache;
+
+  if (cache.has(tx_and_mixring_hash))
+    return true;
+
+  // We had a cache miss, so now we must expand the mix ring and do full verification
+  return expand_tx_and_ver_rct_non_sem(tx, tx_prefix_hash, mix_ring);
```

Comment:
Wait a minute, I don't see where you add successful results to the cache? Where is `cache.add`?

Reply:
That's a great point. Removed it for debugging and forgot to add it back in haha. Check latest commits